python_modules/dagster/dagster/_core/definitions/data_time.py

```diff
@@ -502,6 +504,29 @@ def get_current_data_time(
 
         return min(cast(AbstractSet[datetime.datetime], data_times), default=None)
 
+    def _get_source_data_time(
+        self, asset_key: AssetKey, current_time: datetime.datetime
+    ) -> Optional[datetime.datetime]:
+        latest_record = self.instance_queryer.get_latest_materialization_or_observation_record(
+            AssetKeyPartitionKey(asset_key)
+        )
+        if latest_record is None:
+            return None
+        if latest_record.asset_observation is None:
+            raise DagsterInvariantViolationError(
+                "Cannot calculate data time for source asset with materialization"
```

Comment:
no action required, but `get_latest_materialization_or_observation_record` is a confusing name and does not actually return the latest record of either type for a given key. it will only ever return a materialization event for a non-source asset, and only ever return an observation event for a source asset, so it is not possible for latest_record.asset_observation to be None here (in theory).

doesn't hurt to have the assert though.

Reply:
👁️👁️ . going to use a `check.invariant` in that case